### ISSUE TITLE
Question about Figure 1 in paper

### ISSUE THREAD

**Issue by BooguTeam** [IBM/BadDiffusion#10]
Please correct me if I'm wrong.

In the backdoor diffusion model of training phase in Figure 1, the pic before the forward diffusion should be 
$\sqrt{\alpha_0} \times x\prime_0 + \sqrt{1 - \alpha_0} \times r\prime$, since $\alpha_0$ is much better than $1 - \alpha$, $x\prime_0$ should play a dominant role. So I think the pic should more closer to your target pic not the clean  pic with a trigger.

**Comment by FrankCCCCC** [comment]
Hi, thank you for your valuable advice. I guess you may want to say $`\mathbf{x}_{0}' = \sqrt{\alpha_{0}} \mathbf{x}_{0}' + (1 - \sqrt{\alpha_{0}}) \mathbf{r}`$ is weird. Mathematically, you're right, but empirically, this euqation would hold most of time. The reason is that DDPM sets $`\alpha_{0} = 0`$ and $`\alpha_{T} = 1`$ originally, and thus, the equation holds. Even though some people would choose $`\alpha_{0} \approx 0`$ or $`\alpha_{T} \approx 1`$ for some other reasons, the euqation still holds approximately. If they don't choose the values close to 0 or 1, the diffusion models don't work. I look forward to discussing further with you.

**Comment by Harr7y** [comment]
Thanks for your reply. I have a different understanding for $\alpha_0$ in DDPM, I think it equals to 1 or close to 1, not 0. 

The backdoor in your paper is adding noise with a trigger to the target image for forward diffusion, so I think the pic should more closer to your target pic not the clean pic with a trigger. 
In your code p_losses_diffuser function of "loss.py":
`x_noisy, target = q_sample_diffuser(noise_sched=noise_sched, x_start=x_start, R=R, timesteps=timesteps, noise=noise)`
The "x_start" is the target image (batch["target"]) not a clean image with a trigger.

I hope that didn't confuse you. :joy:

**Comment by FrankCCCCC** [comment]
Sorry for my fault. $`\alpha_{0} \approx 1, \alpha_{T} \approx 0`$ is correct.

On the other hand, I cannot understand what you mean "the pic should more closer to your target pic not the clean pic with a trigger". What does "the pic" mean? I think $`\mathbf{x}_{T}' = \mathbf{r}, \mathbf{x}_{0}' = \mathbf{x}_{0}', \mathbf{r} := \mathbf{x}_{0} \oplus \mathbf{g}`$ is reasonable due to $`\mathbf{x}_{t}' := \sqrt{\bar{\alpha}_{t}} \mathbf{x}_{0}' +  (1 - \sqrt{\bar{\alpha}_{t}}) \mathbf{r} = \sqrt{\alpha}_{t} \mathbf{x}_{t-1}' +  (1 - \sqrt{\alpha}_{t}) \mathbf{r}`$.

Thank you for your interest~

**Comment by Harr7y** [comment]
The pic means the "the pic before the forward diffusion in Figure 1, a woman wears a glasses". I think it is the $x_0\prime$ in your paper, it should be a backdoor target picture (the cat) not the **r**.

**Comment by FrankCCCCC** [comment]
Hello, actually, I think the code is correct. Also, the implementation is consistent with the paper. As a result, I poison the dataset in the dataloader. Then, augment the Gaussian noise and compute the input and correct ouput in the function ``q_sample_diffuser``. 

Thank you very much. Let me know if you have further questions.
